File: D-MASH PWA/not_messenger/js/node_registration_v4.js
'use strict';
// Pure transcript/validation contract. No route or mailbox authority is granted.
(function(global) {
    const hex = bytes => Array.from(bytes, b => b.toString(16).padStart(2,'0')).join('');
    const field = (value, length) => {
        if (typeof value !== 'string' || !(new RegExp('^[0-9a-f]{'+length+'}$')).test(value)) throw Error('Invalid registration field');
        return value;
    };
    const exact = (value, keys) => value && typeof value === 'object' && !Array.isArray(value) && Object.keys(value).sort().join(',') === keys;
    function resource(issuer, recipient, dnss, transcriptHash) {
        field(issuer,64); field(recipient,64); field(dnss,32);
        if (issuer === recipient || !(transcriptHash instanceof Uint8Array) || transcriptHash.length !== 32) throw Error('Invalid registration context');
        return ['D-MASH','NODE-DNSS','V4',issuer,recipient,dnss,hex(transcriptHash)].join('|');
    }
    function verifyRegistration(session, request, difficulty, {now = Math.floor(Date.now()/1000), expectedDnss = null} = {}) {
        try {
            if (session.version !== 4 || session.closed || session.localRole !== 'NODE' || session.peerRole !== 'NODE' ||
                !Number.isInteger(difficulty) || difficulty < 20 || difficulty > 24 || !Number.isSafeInteger(now)) return false;
            if (!exact(request,'dnss,pow,type') || request.type !== 'NODE_REGISTER') return false;
            field(request.dnss,32);
            if (expectedDnss !== null && request.dnss !== field(expectedDnss,32)) return false;
            const work = resource(session.peerId,session.localId,request.dnss,session.transcriptHash);
            const p = request.pow;
            if (!exact(p,'difficulty,digest,expires_at,nonce,resource,type,v') || p.v !== 1 || p.type !== 'DNSS' || p.resource !== work ||
                p.difficulty !== difficulty || !Number.isSafeInteger(p.expires_at) || p.expires_at <= now || p.expires_at > now+180 ||
                !Number.isSafeInteger(p.nonce) || p.nonce < 0) return false;
            field(p.digest,64);
            const digest = global.DmashResourcePow.activationDigest(session.localId,'DNSS',session.peerId,work,p.nonce,p.expires_at);
            return hex(digest) === p.digest && global.DmashResourcePow.leadingZeroBits(digest) >= difficulty;
        } catch (_) { return false; }
    }
    global.DmashNodeRegistrationV4 = Object.freeze({resource,verifyRegistration});
    if (typeof module !== 'undefined') module.exports = global.DmashNodeRegistrationV4;
})(typeof window !== 'undefined' ? window : globalThis);

File: D-MASH/client/backend/node_registration_v4.py
"""Shared v4 directional registration transcript and proof validation.

No network endpoint is enabled by this module. Admission (including Node identity
work/password policy), durable relationship storage and grant issuance are separate
steps; passing this proof never grants mailbox or route ownership.
"""
import re
import time

if __package__:
    from .resource_pow import MIN_ACTIVATION_DIFFICULTY, MAX_ACTIVATION_DIFFICULTY, verify_activation_pow
else:
    from resource_pow import MIN_ACTIVATION_DIFFICULTY, MAX_ACTIVATION_DIFFICULTY, verify_activation_pow


def _hex(value, length):
    if not isinstance(value, str) or not re.fullmatch('[0-9a-f]{' + str(length) + '}', value):
        raise ValueError('invalid registration field')
    return value


def resource(issuer, recipient, dnss, transcript_hash):
    """Issuer pays for its outbound relationship at recipient, independently of reverse."""
    _hex(issuer, 64); _hex(recipient, 64); _hex(dnss, 32)
    if issuer == recipient:
        raise ValueError('self relationship forbidden')
    if not isinstance(transcript_hash, bytes) or len(transcript_hash) != 32:
        raise ValueError('invalid session transcript')
    return '|'.join(('D-MASH', 'NODE-DNSS', 'V4', issuer, recipient, dnss, transcript_hash.hex()))


def verify_registration(session, request, difficulty, *, now=None, expected_dnss=None):
    """Verify ONLY inbound work, using identities authenticated by the handshake.

    The caller must persist/compare the DNSS before sending acceptance, then bind
    the accepted direction to this exact session. Never reuse acceptance on reconnect.
    """
    current = int(time.time()) if now is None else now
    if (session.version != 4 or session.closed or session.local_role != 'NODE'
            or session.peer_role != 'NODE' or type(difficulty) is not int
            or not MIN_ACTIVATION_DIFFICULTY <= difficulty <= MAX_ACTIVATION_DIFFICULTY
            or type(current) is not int):
        return False
    try:
        if not isinstance(request, dict) or set(request) != {'type', 'dnss', 'pow'} or request['type'] != 'NODE_REGISTER':
            return False
        dnss = _hex(request['dnss'], 32)
        if expected_dnss is not None and dnss != _hex(expected_dnss, 32):
            return False
        work = resource(session.peer_id, session.local_id, dnss, session.transcript_hash)
        proof = request['pow']
        if not isinstance(proof, dict) or set(proof) != {'v', 'type', 'resource', 'nonce', 'expires_at', 'difficulty', 'digest'}:
            return False
        if (type(proof['v']) is not int or proof['v'] != 1 or proof['type'] != 'DNSS'
                or proof['resource'] != work or type(proof['difficulty']) is not int
                or proof['difficulty'] != difficulty or type(proof['expires_at']) is not int
                or not current < proof['expires_at'] <= current + 180
                or type(proof['nonce']) is not int or not 0 <= proof['nonce'] <= 2**53-1):
            return False
        _hex(proof['digest'], 64)
        return verify_activation_pow(session.local_id, 'DNSS', session.peer_id, work,
                                     proof['nonce'], proof['expires_at'], difficulty, proof['digest'], now=current)
    except (ValueError, TypeError, AttributeError):
        return False
